python/cuda_parallel/cuda/parallel/experimental/struct.py

```diff
@@ -0,0 +1,133 @@
+from dataclasses import dataclass
+from dataclasses import fields as dataclass_fields
+from typing import Type
+
+import numba
+import numpy as np
+from numba.core import cgutils
+from numba.core.extending import (
+    make_attribute_wrapper,
+    models,
+    register_model,
+    typeof_impl,
+)
+from numba.core.typing import signature as nb_signature
+from numba.core.typing.templates import AttributeTemplate, ConcreteTemplate
+from numba.cuda.cudadecl import registry as cuda_registry
+from numba.extending import lower_builtin
+
+from .typing import GpuStruct
+
+
+def gpu_struct(this: type) -> Type[GpuStruct]:
```

Comment:
Note to reviewers: A lot of the numba typing/lowering stuff going on here is based on https://github.com/gmarkall/numba-accelerated-udfs/blob/main/filigree/numba_extension.py.

Reply:
btw this implementation is deja vu to me as I recall seeing Warp doing similar things... 😄 
https://github.com/NVIDIA/warp/blob/cc3b17b83e7641a140295f033da2141011af103c/warp/codegen.py#L409

Reply:
Wow, yeah that is strikingly similar :)